Auth phase 1 draft (#167)
Kuadrant/mcp-gateway: docs/design/auth.md

## Auth with the MCP Gateway (Phase 1)

> To provide auth intetgration, the MCP gateway uses (Kuadrant)[https://kuadrant.io] and Gateway API in combination with Keycloak as an identity provider.

Reviewer: ```suggestion
To provide auth integration, the MCP gateway uses (Kuadrant)[https://kuadrant.io] and theGateway API in combination with Keycloak as an identity provider.
```

We are bringing these components together to provide a reference implementation of the OAuth 2.0 Protected Resource Metadata as per the (MCP Authorization)[https://modelcontextprotocol.io/specification/draft/basic/authorization]  and a reference implementation for authorizing access to particular tool.

> Note while we use keycloak in our examples, in theory other IDPs could also be used.

### Authentication Implementation

![](./images/MCPgateway-auth.jpg)

### Protecting the MCP Gateway HTTP endpoint(s)

> In order to enforce authenticated access, the gateway listeners exposing the MCP gateway to clients, can use a kuadrant `AuthPolicy` resource. This resource can define what authentication must be met and also enforce Authz requirements at the gateway (Authorization Covered Below).

Reviewer: ```suggestion
In order to enforce authenticated access, the gateway listeners exposing the MCP gateway to clients, can use a Kuadrant `AuthPolicy` resource. This resource can define what authentication must be met and also enforce AuthZ requirements at the gateway (Authorization covered below).
```

[Example AuthPolicy](./../../config/mcp-system/authpolicy.yaml)

[Example Oauth Setup](../../README.md#example-oauth-setup)

> With this policy in place any unauthenticated request not going to the /.well-known endpoints will get a 401 with a resource_metadata url specified. It is then up to the client to leverage this information to retrieve the resource metadata to know how to authenticate to access the resource (example resource would be /mcp). This resource metadata will be served via the MCP Broker component. 

Reviewer 2: > This will make all requests to go through whenever it passes authentication

yeah, this was very much intentional to just do an authN check (no authZ) with that example.
We'll need to evolve the expression for authZ when we work through a good example e2e with keycloak where scopes ultimately map to some permissions on tool calls.

> Authorino will automatically inject the authentication challenge header in the following format

Am I reading into this too much, or does authorino have some built in support for https://datatracker.ietf.org/doc/html/rfc9728#name-www-authenticate-response to satisfy https://modelcontextprotocol.io/specification/2025-06-18/basic/authorization#authorization-server-location, 
so you wouldn't have to add something like this to the AuthPolicy?

```yaml
        unauthenticated:
          code: 401
          headers:
            'WWW-Authenticate':
              value: Bearer resource_metadata=http://mcp.127-0-0-1.sslip.io:8888/.well-known/oauth-protected-resource/mcp
          body:
            value: |
              {
                "error": "Forbidden",
                "message": "Access denied."
              }
```

> With this policy in place any unauthenticated request not going to the /.well-known endpoints will get a 401 with a resource_metadata url specified. It is then up to the client to leverage this information to retrieve the resource metadata to know how to authenticate to access the resource (example resource would be /mcp). This resource metadata will be served via the MCP Broker component. 

Reviewer: ```suggestion
With this policy in place any unauthenticated request not going to the /.well-known endpoints will get a 401 with a resource_metadata URL specified. It is then up to the client to leverage this information to retrieve the resource metadata to know how to authenticate to access the resource (example resource would be /mcp). This resource metadata will be served via the MCP Broker component. 
```

To understand the full flow including the defined Auth server take a look at [MCP Gateway Auth Sequence Diagram](./flows.md#mcp-gateway-request-authentication)

### Registering required scopes

As the MCP gateway is aggregating multiple MCP Servers together. We need to know what scopes are required when asking for an authorization token from the auth provider.  To do this each MCP server registered via the `MCPServer` resource that targets a HTTPRoute at the Gateway will declare its required scopes:

```yaml

## not actual implmentation may be slightly different.

name: accounting_mcp
  spec:
    # Reference all three test MCP servers via their HTTPRoutes
    targetRefs:
    # Server 1 - Go SDK based (hi, time, slow, headers tools)
    - group: gateway.networking.k8s.io
      kind: HTTPRoute
      name: mcp-server1-route
    toolPrefix: s3_
    auth: #just an example that may work
      scopes: 
        - email
        - role
        - user
      rbac:
        - role: accounting
          tools: 
            - "payroll"
        - role: admin
          tools: 

>             -"*"

Reviewer 3: If this info can be compiled dynamically from multiple (hypothetical) config files of each tool, then perhaps we want here some kind of selector of the tools/config files themselves instead, as opposed to duplicating the info into the MCPServer resource?

On the other hand, if it's a static list (that occasionally may diverge from the actual tools running/being served behind), then I wonder if the info doesn't belong into the AuthPolicy.

In fact, why having an `auth` section in the `MCPServer` resource et all? Couldn't this (shouldn't this?) all be in the AuthPolicy? Unless we're thinking on hiding the creation/ownership of the AuthPolicy resource behind a MCPServer one as a high level API?

Author: So we have 2 options we can show. One the ACL server and 2 via keycloak. I am moving tool filtering to phase 2 doc

```

For phase1 this set of scopes will be aggreggated together via the MCP Server controller into the config for the MCP Broker. The MCP broker is then responsible for serving the aggregated scopes via `./well-known/oauth-protected-resource` endpoint. 

```json
{
  "resource_name":"MCP Server",
  "resource":"http://mcp.127-0-0-1.sslip.io:8888/mcp",
  "authorization_servers":["http://keycloak.127-0-0-1.sslip.io:8888/realms/mcp"],
  "bearer_methods_supported": ["header"],
  "scopes_supported":["email","role","user"]
}

```

> Note In a future phase these defined scopes may be used to define a token exchange to scope down the token to only the required scopes. 

### Registering MCP Auth Severs

> Part of the response from the `oauth-protected-resource` endpoint is an array of `authorization_servers`. For phase 1 the gateway will only support a single auth server. The expectation is each of the MCPServers using OAuth is using the same auth server and will independantly validate the token against that auth server. The MCP Broker will be configured with this shared server separately via a flag or EnvVar.

Reviewer: ```suggestion
Part of the response from the `oauth-protected-resource` endpoint is an array of `authorization_servers`. For phase 1 the gateway will only support a single auth server. The expectation is that each of the MCPServers using OAuth are using the same auth server and will independently validate the token against that auth server. The MCP Broker will be configured with this shared server separately via a flag or EnvVar.
```

> > Note this doesn't include URL elicitation where an MCP server may request a user provide it access to a resource not protected by the defined auth server. Example a github resource/API. This is consisdered seaparate from the general Gateway authentication and authorization requirement.

Reviewer 3: This could be achieved using [**Keycloak Authorization Services**](https://www.keycloak.org/docs/latest/authorization_services/index.html).

Example in an AuthConfig/AuthPolicy: https://github.com/Kuadrant/authorino/blob/main/docs/user-guides/keycloak-authorization-services.md

Reviewer 2: Just confirming, your comment is about the elicitation text, not the text above that for the `oauth-protected-resource` endpoint?

For elicitation (or avoiding it), if I'm understanding what you're suggesting, there would be backchannel communication between authorino and the 3rd party server to request and get some level of access on behalf of the user, but without the user (or the client) being part of that flow?

Reviewer 3: Yes, the elicitation part.

In the example I shared (Authorino user guide), the backchannel is for Authorino to open a ticket with the auth server, on behalf of the requestor, for the owner of the resource to decide on authorisation. The owner of the resource still needs to go to the Keycloak UI and grant/deny the requested access. If access is granted, then the request can be repeated and it goes through.

There are a few caveats. One is that the resource needs to be tracked by the auth server, with the proper ownership association stored there. From my experience, that's easily said than done, though not impossible.

Another caveat, at least with the example I shared, is that there's no support for Keycloak Authorization Services built-in in a Kuadrant AuthPolicy or Authorino AuthConfig today. The example was crafted using OPA, leaking some client secrets into the spec of the policy. That's not ideal.

### Authenticated Calls

Once a client has obtained a token, it can then make requests to the MCP Gateway. When a request comes to the gateway, the Kuadrant WASM plugin intercepts this request and based on configuration, will decide whether or not to call to the Authorino component. With the linked AuthPolicy, Authorino will receive the request and then validate the token with the configured auth server before allowing the request to continue. 

> Note not covered here is message signinging as specified under https://github.com/modelcontextprotocol/modelcontextprotocol/issues/1415 . This will be covered at a later phase. 

### Authorization of tools/calls

> When the router receives a tools/call. It will load from the config the RBAC section for the specified MCP server. It will then set this data into the dynamic metadata via the `ext_auth_data` namespace. Optionally we could also encode this data and set it as a request header. The reason for this is to allow downstream filters to access this information to make authorization decisions. 

Author: I have done this now. AuthPolicy requests this metadata from a separate "server" in the example it is just a static file but obv could be anything that Authorino can interact with https://github.com/kagenti/mcp-gateway/pull/196/files#diff-80c5f77bc4ee52cd5426f9b1fda44c143c58084313203dc47354e59ecc5cddbeR17

To do this we will update the envoy filter to allow it define metadata: https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_proc/v3/ext_proc.proto#envoy-v3-api-msg-extensions-filters-http-ext-proc-v3-metadataoptions 

```yaml
metadata_options:
  receiving_namespaces:
    untyped:
      - ext_auth_data
```                  

Note> We will use AuthPolicy to enforce the tool access check as a reference of how you can achieve this integration with the gateway and Kuadrant. 

### Filtering of tools

> The broker, when receiving a tools/list call, will validate the JWT token if present in the request, and load the RBAC data from the config. As it has access to the JWT it can identify which tools should be returned based on this configuration. 

Reviewer 2: So this comes down to if the filtering of the tool list should be part of the 'mcp gateway',
or should the 'mcp gateway' support the ability to allow something to filter that list response per user.

The allow/deny of an individual tool call per user feels like its firmly in the realm of an auth integration that the mcp-gateway supports, not part of the mcp-gateway itself.
However, when filtering the response body, it's a little different than an allow/deny filter. That's where I'm hesitating. 
I'm also aware of not introducing hard dependency on kuadrant for any of the basic mcp-gateway functionality.
So an important question here is if filtering of the tool list response is basic functionality.
If not, then how do we allow plugging in something that can do that filtering (including kuadrant as an example) in a decoupled manner.

> The broker, when receiving a tools/list call, will validate the JWT token if present in the request, and load the RBAC data from the config. As it has access to the JWT it can identify which tools should be returned based on this configuration. 

Author: going to remove tool filtering from this and move to phase 2

## Auth with the MCP Gateway (Phase 1)

> To provide auth integration, the MCP gateway uses (Kuadrant)[https://kuadrant.io] and theGateway API in combination with Keycloak as an identity provider.

Reviewer: ```suggestion
To provide auth integration, the MCP gateway uses (Kuadrant)[https://kuadrant.io] and the Gateway API in combination with Keycloak as an identity provider.
```
